wallet/src/account/output_cache/mod.rs

```diff
@@ -60,35 +59,50 @@ impl OutputCache {
         cache
     }
 
-    pub fn txs(&self) -> &BTreeMap<AccountWalletTxId, WalletTx> {
-        &self.txs
+    pub fn txs_with_unconfirmed(
+        &self,
+    ) -> impl Iterator<Item = (&OutPointSourceId, &WalletTx)> + '_ {
+        self.txs.iter()
     }
 
-    pub fn outpoints(&self) -> &BTreeSet<UtxoOutPoint> {
-        &self.consumed
+    pub fn get_txo(&self, outpoint: &UtxoOutPoint) -> Option<&TxOutput> {
+        self.txs
+            .get(&outpoint.tx_id())
+            .and_then(|tx| tx.outputs().get(outpoint.output_index() as usize))
     }
 
     pub fn add_tx(&mut self, tx_id: AccountWalletTxId, tx: WalletTx) {
+        let is_unconfirmed = match tx.state() {
+            TxState::InMempool => true,
+            TxState::Confirmed(_, _) | TxState::Conflicted(_) | TxState::Inactive => false,
+        };
+
         for input in tx.inputs() {
             match input {
                 TxInput::Utxo(outpoint) => {
-                    self.consumed.insert(outpoint.clone());
+                    if is_unconfirmed {
+                        self.unconfirmed_consumed.insert(outpoint.clone());
+                    } else {
+                        self.consumed.insert(outpoint.clone());
+                    }
                 }
                 TxInput::Account(_) => {
                     unimplemented!()
                 }
             }
         }
-        self.txs.insert(tx_id, tx);
+
+        self.txs.insert(tx_id.into_item_id(), tx);
     }
 
     pub fn remove_tx(&mut self, tx_id: &AccountWalletTxId) {
-        let tx_opt = self.txs.remove(tx_id);
+        let tx_opt = self.txs.remove(tx_id.item_id());
         if let Some(tx) = tx_opt {
             for input in tx.inputs() {
                 match input {
                     TxInput::Utxo(outpoint) => {
                         self.consumed.remove(outpoint);
+                        self.unconfirmed_consumed.remove(outpoint);
                     }
```

Comment:
Maybe overkill, but instead of calling remove() on both, should this also do `match tx.state()` here and then call the right one? May not matter now, but just thinking in the future if more states would ever be added?

Reply:
One thing I am thinking now is to maybe change it from 2 x Set<UtxoOutPoint> to a Map<UtxoOutPoint, TxState>, where the value is the TxState of the transaction that consumed it.